### ISSUE TITLE
Document Use of Compilers Requiring Modules

### ISSUE THREAD

**Issue by citibeth** [spack/spack#11582]
Some compilers (eg Intel) require modules to load.  Traditionally, they can only be used by loading the required modules and then using `spack --dirty`.  Apparently, there's another way to use them, an "rpath trick."  See below.  This needs to be added to the documentation.

On Wed, May 29, 2019 at 11:44 AM 'Kenneth Moreland' via Spack <spack@googlegroups.com> wrote:
Bili,

Thanks for the tip. The setting of extra_rpaths seems to do the trick. As a reference for anyone else who runs into this, here is the relevant  state of my compilers.yaml file:

```
compilers:
- compiler:
    environment: {}
    extra_rpaths:
    - /opt/intel/compilers_and_libraries_2018.2.199/linux/compiler/lib/intel64_lin
    flags: {}
    modules:
    - intel/18.0.2
    operating_system: centos7
    paths:
      cc: /opt/intel/compilers_and_libraries_2018.2.199/linux/bin/intel64/icc
      cxx: /opt/intel/compilers_and_libraries_2018.2.199/linux/bin/intel64/icpc
      f77: /opt/intel/compilers_and_libraries_2018.2.199/linux/bin/intel64/ifort
      fc: /opt/intel/compilers_and_libraries_2018.2.199/linux/bin/intel64/ifort
    spec: intel@18.0.2
    target: x86_64
```

I also agree with Fabricio that it would be helpful if Spack automatically added these rpaths to compilers that it finds.

-Ken

On Wednesday, May 29, 2019 at 6:46:57 AM UTC-6, Fabricio Cannini wrote:
I'd second the extra_rpath trick.
I've had similar problems with externally installed gcc and fixed it the same way.

I'd go further and suggest that "spack compiler find" should set this option by default for all externally installed compilers. <wink wink nudge>

[ ]'s

Em qua, 29 de mai de 2019 02:33, Bili Dong <qobi...@gmail.com> escreveu:
Hi Ken,

I just reproduced this issue on Stampede2 and found this extra_rpath trick to be the solution. You'll need to reinstall python with the new compilers.yaml and the issue shall go away.

Best,
Bili

On Tuesday, May 28, 2019 at 9:21:50 PM UTC-7, Kenneth Moreland wrote:
I am in the process of trying to compile some code on Stampede2 through Spack with the Intel compiler. For the most part I have my settings working to compile with the Intel compiler. However, one of the (deep) dependencies is using some Python utilities, and when Spack tries to install the py-setuptools package, it fails to run the internally compiled python with the following error.

==> Installing py-setuptools
==> Searching for binary cache of py-setuptools
==> Warning: No Spack mirrors are currently configured
==> No binary for py-setuptools found: installing from source
/home1/06485/kmorel/spack/opt/spack/linux-centos7-x86_64/intel-18.0.2/python-2.7.16-3nshxebi6ftvii7inidep7d6ox6gre2d/bin/python2.7: error while loading shared libraries: libimf.so: cannot open shared object file: No such file or directory
==> Error: ProcessError: Command exited with status 127:
    '/home1/06485/kmorel/spack/opt/spack/linux-centos7-x86_64/intel-18.0.2/python-2.7.16-3nshxebi6ftvii7inidep7d6ox6gre2d/bin/python2.7' '-c' 'from distutils.sysconfig import get_config_var; print get_config_var('"'"'LIBDIR'"'"')'

It looks like the python2.7 compiled with the Intel compiler cannot be run when installing the py-setuptools package although it seems to work when I run it from the command line.

What I think is happening is that executables compiled with the Intel compiler still need the Intel compiler module loaded. (If I unload the module in my shell and try to run python again from the command line, I get the same error.) For most packages this is not an issue since most require some form of a compiler and will load the Intel compiler module. But py-setuptools (presumably) does not use any compiler and so does not load a module.

So, I think the problem would be solved if all packages loaded the Intel compiler package/module regardless of whether it was compiling anything. Is there any way to set the Spack configuration to always do that?

-Ken
-- 
You received this message because you are subscribed to the Google Groups "Spack" group.
To unsubscribe from this group and stop receiving emails from it, send an email to sp...@googlegroups.com.
To post to this group, send email to sp...@googlegroups.com.
Visit this group at https://groups.google.com/group/spack.
To view this discussion on the web visit https://groups.google.com/d/msgid/spack/2e82c5ac-b102-48d3-9fa9-29e1268aebb7%40googlegroups.com.
For more options, visit https://groups.google.com/d/optout.

Em qua, 29 de mai de 2019 02:33, Bili Dong <qobi...@gmail.com> escreveu:
Hi Ken,

I just reproduced this issue on Stampede2 and found this extra_rpath trick to be the solution. You'll need to reinstall python with the new compilers.yaml and the issue shall go away.

Best,
Bili

On Tuesday, May 28, 2019 at 9:21:50 PM UTC-7, Kenneth Moreland wrote:
I am in the process of trying to compile some code on Stampede2 through Spack with the Intel compiler. For the most part I have my settings working to compile with the Intel compiler. However, one of the (deep) dependencies is using some Python utilities, and when Spack tries to install the py-setuptools package, it fails to run the internally compiled python with the following error.

==> Installing py-setuptools
==> Searching for binary cache of py-setuptools
==> Warning: No Spack mirrors are currently configured
==> No binary for py-setuptools found: installing from source
/home1/06485/kmorel/spack/opt/spack/linux-centos7-x86_64/intel-18.0.2/python-2.7.16-3nshxebi6ftvii7inidep7d6ox6gre2d/bin/python2.7: error while loading shared libraries: libimf.so: cannot open shared object file: No such file or directory
==> Error: ProcessError: Command exited with status 127:
    '/home1/06485/kmorel/spack/opt/spack/linux-centos7-x86_64/intel-18.0.2/python-2.7.16-3nshxebi6ftvii7inidep7d6ox6gre2d/bin/python2.7' '-c' 'from distutils.sysconfig import get_config_var; print get_config_var('"'"'LIBDIR'"'"')'

It looks like the python2.7 compiled with the Intel compiler cannot be run when installing the py-setuptools package although it seems to work when I run it from the command line.

What I think is happening is that executables compiled with the Intel compiler still need the Intel compiler module loaded. (If I unload the module in my shell and try to run python again from the command line, I get the same error.) For most packages this is not an issue since most require some form of a compiler and will load the Intel compiler module. But py-setuptools (presumably) does not use any compiler and so does not load a module.

So, I think the problem would be solved if all packages loaded the Intel compiler package/module regardless of whether it was compiling anything. Is there any way to set the Spack configuration to always do that?

-Ken
-- 
You received this message because you are subscribed to the Google Groups "Spack" group.
To unsubscribe from this group and stop receiving emails from it, send an email to sp...@googlegroups.com.
To post to this group, send email to sp...@googlegroups.com.
Visit this group at https://groups.google.com/group/spack.
To view this discussion on the web visit https://groups.google.com/d/msgid/spack/2e82c5ac-b102-48d3-9fa9-29e1268aebb7%40googlegroups.com.
For more options, visit https://groups.google.com/d/optout.
-- 
You received this message because you are subscribed to the Google Groups "Spack" group.
To unsubscribe from this group and stop receiving emails from it, send an email to spack+unsubscribe@googlegroups.com.
To post to this group, send email to spack@googlegroups.com.
Visit this group at https://groups.google.com/group/spack.
To view this discussion on the web visit https://groups.google.com/d/msgid/spack/34941e20-9999-4158-a69e-9617442c8eb0%40googlegroups.com.
For more options, visit https://groups.google.com/d/optout.

**Comment by citibeth** [comment]
I dunno.... Spack-build and system compilers don't need this.

**Comment by adamjstewart** [comment]
> Some compilers (eg Intel) require modules to load. Traditionally, they can only be used by loading the required modules and then using `spack --dirty`.

You can also specify which modules need to be loaded inside `compilers.yaml` and Spack will automatically load them during build time.

**Comment by citibeth** [comment]
That didn't work for me, I forget why.  I think the problem was it only
allowed one module to be loaded, but I needed multiple.

**Comment by kmorel** [comment]
> You can also specify which modules need to be loaded inside `compilers.yaml` and Spack will automatically load them during build time.

That works for me but _only if the package being built uses the compiler_. In the issue I raised, the Intel compiler module was used to build the `python` package, and then the subsequent `py-setuptools` package tried to run the python executable. But because `py-setuptools` doesn't use the Intel compiler, it doesn't load the Intel compiler module, and thus fails.

**Comment by mwkrentel** [comment]
I asked about this 8 months ago in #9376 and @floquet asked about this
3 months ago in #10790, but neither of us got any traction.

My point is: even in the simplest possible case, there already is a
working compiler module and `spack compiler find` finds it, then there
is still an extra manual step to fill in the `modules:` field in
`compilers.yaml`.  It's just too easy to make that mistake.

Now, you're talking about needing multiple modules (often true for
Intel and Cray) and adding rpaths.  Even if spack has enough machinery
in `compilers.yaml` to make this work without `--dirty`, then it's
still a manual step and a very easy mistake to make.

I'm not claiming this is easy, but anything spack can do to automate
the process will help save a lot of user errors (and "this doesn't
build" issues).

Even a simple message from `compiler find` that says, "You just added
a compiler from a module, please check the following fields and add
them to `compilers.yaml`."

**Comment by adamjstewart** [comment]
@svenevs may also be interested in this conversation. I know he has to load the GCC module before loading the Intel module to get his setup working.

**Comment by svenevs** [comment]
In my opinion spack should do nothing with autopopulation. In particular, rpath is only one part of the puzzle. Include paths are another. And there are others most likely. If spack tries to do this automatically, chances are good it won't be correct all of the time. The spack generated modules take care of everything already. In particular, if you look at the Intel module, there's a lot that needs to be discovered spread across many directories.

Put differently: I believe spack should take the stance of "the user is responsible for configuring their complier modules correctly". I have very serious doubts about being able to autopopulate everything correctly, and would probably cry if somebody asked me to implement it. There are too many compilers with too many quirks, and the quirks often change across versions.

I've tried documenting this before and it went nowhere. I'm willing to redo the docs more concisely if

1. People agree that simply documenting the importance of loading modules is the right solution. Examples will be Intel (with or without custom GCC, since system GCC may not be supported) and spack installed llvm.
2. I am told exactly where on which document this should be added.

Basically, I'm happy to document this because it needs to be explicitly explained to users. But I have an insane workload right now and don't want to bother unless it will actually get merged.

**Comment by svenevs** [comment]
@citibeth multiple compiler modules can be loaded. It sounds like the solution to your python problem would be to make sure that modules are always loaded in the environment. I haven't experienced your problem, but that sounds like it would fix it?

**Comment by kmorel** [comment]
> It sounds like the solution to your python problem would be to make sure that modules are always loaded in the environment. I haven't experienced your problem, but that sounds like it would fix it?

It sounds like it would fix my original problem.

> I believe spack should take the stance of "the user is responsible for configuring their complier modules correctly". I have very serious doubts about being able to autopopulate everything correctly, and would probably cry if somebody asked me to implement it.

I can respect that it is too hard to try to get the compiler configuration perfect for every compiler on every system that Spack might be run on. So the simpler solution of having the compiler module loaded for all packages sounds reasonable to me.

That said, as a Spack newbie I'd like to point to a separate issue (that should probably be in a different issue report) that it would be nice if Spack was better at populating the `modules` field in the compilers.yaml configuration. Since Spack has no way of establishing the full environment needed to leverage compilers, then it should lean on the modules provided by the system administration to set up the environment for Spack. At the very least, it would be nice to have a `spack compiler find module` to autopopulate compiler configuration that loads the given module.

**Comment by adamjstewart** [comment]
I believe Spack already autopopulates the modules field on Cray. It shouldn't be hard to do the same on non-Cray systems. If a compiler is found via a module file, we should add the module. If it is found via PATH, we should add the path.

**Comment by citibeth** [comment]
> @citibeth multiple compiler modules can be loaded. It sounds like the solution to your python problem would be to make sure that modules are always loaded in the environment. I haven't experienced your problem, but that sounds like it would fix it?

Yes, I think I had the "python problem."  Remember that whatever problems I had at the time might have been fixed by now.  But whatever the issue... I think the best would be for a Spack Environment to specify modules that should always be loaded when building and running that environment.  My modules weren't just for the compiler, but also for MPI and a couple of other things.

**Comment by mwkrentel** [comment]
Yes, Cray is special for filling in compiler modules, in part due to
the heavy use of modules and PrgEnv.  But only Cray.  Everywhere else
requires a manual step.

**Comment by svenevs** [comment]
Ok so action item that can be solved immediately: compiler modules are always loaded for every package, not just packages that use the compiler. Not sure how or where that happens, it sounds like the compiler wrappers load the module, and needs to be moved.

For auto finding modules, it's a little more challenging. Right now, you typically `spack load llvm` and then `spack compiler find`. Proposition: `spack compiler find [spec]`, e.g. `spack compiler find llvm`. I believe that having access to the spec would make finding the module easier.

I use lmod, but I'll note that my compiler modules always use tcl version to avoid getting evicted via `family("complier")`. Otherwise when I need to load GCC first and then Intel, it won't work using the lmod modules.

The spec addition also places the responsibility of the user to still populate modules correctly for scenarios that are harder to know exactly which module made something available. For example, different GCC toolchains on RHEL and devtoolset repos. Or an external Intel installation.

**Comment by mwkrentel** [comment]
> For auto finding modules, it's a little more challenging. Right now,
> you typically spack load llvm and then spack compiler
> find. Proposition: spack compiler find [spec], e.g. spack compiler
> find llvm. I believe that having access to the spec would make
> finding the module easier.

Keep in mind that the compiler module doesn't have to be built by
spack.  There are plenty of systems with pre-built modules for
compilers and MPI where I want to use the existing installation.

For external modules, maybe there could be a syntax like
`spack compiler module <name>` where you tell spack that `name`
is the name of a module that contains a compiler.

**Comment by citibeth** [comment]
On Fri, May 31, 2019 at 12:55 PM Stephen McDowell <notifications@github.com>
wrote:

> Ok so action item that can be solved immediately: compiler modules are
> always loaded for every package, not just packages that use the compiler.
>

Then they're not compiler modules (from Spack's perspectives): they're just
modules, or environment-related modules that get loaded for a partiular
environment.

**Comment by citibeth** [comment]
> For external modules, maybe there could be a syntax like
> spack compiler module <name> where you tell spack that name
> is the name of a module that contains a compiler.
>

Or just edit the appropriate YAML file.  We don't need a one-off Spack
command for every possible YAML edit one might want to do.

>

**Comment by chuckatkins** [comment]
Rather than manually adding rpath info, would it be possible to instead convey that the compiler package itself needs to become a downstream run-time dependency of anything it builds?  I experience the same problem using gcc-8 out of the openhpc distro.  I generally work around it by just manually loading the module before I start working on things, since even though the compilers.yaml file has the module, this same issue crops up since runtime library dependencies get added to everything it builds.

Or even more sweeping, should a compiler package *always* be added as a runtime dependency of anything it builds?

**Comment by kmanalo** [comment]
As it relates to the title "Compilers Requiring Modules" and the original comment. I dealt with ongoing issues in building Trilinos with Intel 19 backed by GCC 8 with MVAPICH2 for MPI.  I nearly used the same prescription for `modules.yaml` as the original comment shows, but what I found may be specific to CMake builds.  I had trouble loading a gcc/8.3.0 module with the 'modules' key, and I didn't want to use the `--dirty` option.

I hit the "includes" problem.  My current workaround was to use `CMAKE_MODULE_PATH` to find the includes needed for GCC 8. Somewhere in beast CMake build process there was a `cpp -v` to search include paths, which kept defaulting to the base Intel 19 with gcc 4.8.5 on a RHEL7 system.  

I believe this issue is a bit of 'needs documentation' and 'needs per package handling'.  Stronger detection and reporting in `spack compiler` could be useful.  Despite all the trouble I was having, I align very strongly with the comment by @svenevs "the user is responsible for configuring their complier modules correctly".

**Comment by kmanalo** [comment]
@svenevs comment:
> I use lmod, but I'll note that my compiler modules always use tcl version to avoid getting evicted via `family("compiler")`. Otherwise when I need to load GCC first and then Intel, it won't work using the lmod modules.

@svenevs when using Lmod hierarchy with Lua module files, Lmod allows for `export LMOD_EXPERT=1` to load multiple compilers.

The above may not be easy to explain to users, so at many sites have just added a `gcc-compatibility` module which adds the necessary `PATH`s to allow Intel to get the backing gcc. (then checking with `icpc -v`).

**Comment by tjahns** [comment]
I stumbled across this issue when debugging my own compiler Pokemon hunt:
The [documentation](https://spack.readthedocs.io/en/latest/getting_started.html#configuration-with-modules) states that multiple modules can be specified like this:
`modules: [ mod1, mod2 ]`
but indeed only
`modules: [ mod1 mod2 ]`
where all the modules are combined into one string works.

**Comment by msimberg** [comment]
I've been attempting to use environments on a Cray system (Piz Daint) and I'm struggling to get MPI and compiler modules loaded when activating an environment. Following this thread (and not finding anything regarding this in the documentation) it's still not clear to me if what I'm trying to do is expected behaviour or a bug.

My `packages.yaml`:
```
packages:
  all:
    compiler:
      - gcc
      - intel
      - pgi
      - cce
      providers:
        # snip
        mpi:
          - cray-mpich
          - mpich
          - openmpi
  cray-mpich:
    buildable: false
    externals:
    - modules:
      - cray-mpich/7.7.16
      spec: cray-mpich@7.7.16 %gcc@9.3.0.21.02
    # other versions follow
```
and in `compilers.yaml`
```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-gnu
    - gcc/9.3.0
    - cdt/21.02
    - gcc/9.3.0
    operating_system: cnl7
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: gcc@9.3.0.21.02
    target: any
# other compilers follow
```

I then have an environment file which looks like this:
```
spack:
  view: true
  specs: [mpi%gcc@9.3.0.21.02]
  concretization: together
```

and concretizing it gives me the expected result:
```
==> Concretized mpi%gcc@9.3.0.21.02
 -   wp5zkxb  cray-mpich@7.7.16%gcc@9.3.0.21.02 arch=cray-cnl7-haswell
```

However, no modules are loaded, nor is the compiler available through `CC` or `$CXX`.

It sounds to me from reading this thread that this is expected, but is it? To be clear, if I use `build-env` with a spec that depends on these compiler and package definitions all is good. I'm specifically trying to use `spack env` for a more ad-hoc environment. Please let me know if this is not actually an intended use case for environments. Additionally, if the "rpath trick" is applicable in this case, it's not clear to me what exactly needs to go in there. Finally, I can't comment the contents of the `packages.yaml` and `compilers.yaml` files since I haven't set them up, but if it looks like they are faulty I'd like to upstream any required changes to those who set up our spack configuration.

**Comment by scheibelp** [comment]
@msimberg

> It sounds to me from reading this thread that this is expected, but is it? To be clear, if I use build-env with a spec that depends on these compiler and package definitions all is good.

It sounds like you are defining a compiler in the environment's configuration, activating the environment, and expecting the compiler to be available for use - e.g. in PATH (perhaps you want to use the environment to build some packages without Spack (e.g. you have packages that you build without `spack install`): that is not supported (in general, environments can only "expose" installations of Spack packages - those with a `package.py` file).

For compilers with associated Spack packages (e.g. there is a `gcc` Spack package such that you could say `spack install gcc`) you could in fact make the compiler available with `spack env activate` if you `spack add <the compiler>` for the environment, but I don't think we have a package definition for the Cray compilers.

Also, I think your question is sufficiently different that it should be its own thread. If this answer is not sufficient, could you open a new issue and ping me on it?

**Comment by mwkrentel** [comment]
@msimberg This probably isn't the problem, but in your `packages.yaml` file, you have `providers:`
under `compilers:`.  You probably want it under `all:` instead.

**Comment by msimberg** [comment]
> It sounds like you are defining a compiler in the environment's configuration, activating the environment, and expecting the compiler to be available for use - e.g. in PATH (perhaps you want to use the environment to build some packages without Spack (e.g. you have packages that you build without `spack install`): that is not supported (in general, environments can only "expose" installations of Spack packages - those with a `package.py` file).

Yes, that's what I was expecting, or at least hoping for. This issue sounded close enough to what I was hoping to do but I wasn't sure if it was applicable. This clears it up, thank you!

> Also, I think your question is sufficiently different that it should be its own thread. If this answer is not sufficient, could you open a new issue and ping me on it?

I can certainly open an issue if you think it's useful. I wasn't sure if this behaviour is by design or just because of noone requesting/implementing it yet. I can certainly work with `build-env` if that's the recommended way to do this. I was mostly looking to correct my assumptions about `spack env`.

> @msimberg This probably isn't the problem, but in your `packages.yaml` file, you have `providers:`
> under `compilers:`. You probably want it under `all:` instead.

Sorry for misleading you there. I abbreviated a much longer config file and got the indentation wrong. It is indeed under `all:` in the actual config.

**Comment by germasch** [comment]
I got pointed to this thread, dealing with the same issue, and hoping there's a better solution.

> in general, environments can only "expose" installations of Spack packages - those with a package.py file

IMHO this is a kinda limited definition of an environment. I mean generally you're exposing these installations because you want to do something with them, most likely to actually build some code that depends on those packages, and hence needing matching compilers is probably a rather common case.

So what's the recommend way to deal with this situation -- is setting up my development environment just a multi-step process which consists of a spack environment and then additional manual modules or path modifications (yeah, I know, I can source a script for that)? Or is there an alternate workflow that is recommended? I haven't quite gotten the hang of `spack develop`, but I suppose like `spack build-env` it's for developing spack packages, right, so not really applicable here?

**Comment by scheibelp** [comment]
> I haven't quite gotten the hang of spack develop, but I suppose like spack build-env it's for developing spack packages, right, so not really applicable here?

Yes: if you can't make a Spack `package.py` file then you would need to add your own script to make a given compiler the chosen one for your working environment `spack-python` might be able to help you do that automatically (i.e. add the compiler appropriate for whatever environment you activate). We assume it would be easy to create a Spack package for development purposes but if that doesn't apply in your case, then I'd be interested to hear why.

There are two possible developments that might help (not implemented yet):

* (planned) Spack is planning on treating compilers as normal Spack dependencies and once that is done, it might be exposed automatically (or at least, it should be easy to customize at that point)
* (not planned currently) we could add a section for custom shell modifications to the environment (so e.g. you could add the chosen compiler to `PATH`)

(minor side note about `build-env`: it is also useful when writing a new Spack package for a code you are not necessarily developing, since it allows you to interact with the code using all the shell setup applied by Spack)

**Comment by haampie** [comment]
> it is also useful when writing a new Spack package for a code you are not necessarily developing

I was about to point this out too, particularly considering that loading "the compiler of the environment" is not well-defined since environments contain multiple packages with potentially multiple compilers, and you can't just load all the build environments of all dependencies without conflicts. 

We should just make it easier for people to create packages with development in mind. In particular if you want to share them with others (which is one of the main selling points of environments).

The current workflow for a shareable development package is this long list of steps:

1. cd to a new directory: `cd $(mktemp -d)`
2. create an environment: `spack env create -d .`
3. create a custom repo: `spack repo create repo temp`
4. add the repo to the environment: `spack -e . repo add repo`
5. create a new package in there, e.g.: `spack create -r repo -t cmake my-dummy-package`, do the 6 fixme's, remove the url/homepage, set `has_code = False`, create a fake version, add dependencies...
6. add the package to the environment `spack -e . add my-dummy-package`
7. maybe setup a `packages:` section with externals and concretization preferences
8. run `spack -e . concretize`
9. jump into the build environment: `spack -e . build-env my-dummy-package`

And if you then continue development and add sources and wanna test actual installs:

10. run `spack -e . develop -p ./my_sources my-dummy-package@1.2.3`
11. `spack -e . concretize`
12. `spack -e . install`

I guess the problem is that the focus of creating packages has been around released tarballs from a url, and the focus of environments has been software installs for sites; environments only later got support for development purposes, maybe packages should be improved in this sense too (like can i have an anonymous package or something).

**Comment by germasch** [comment]
[Circling back to the original question about environments, I agree that it's not trivial to know what compiler etc is wanted, and it may not even be well defined. So I don't have a real good answer, but I do think it'd be nicer for stuff to "just work".]

So actually writing my own package files was my first approach, and in fact I'd like that to be my only approach so that I don't have to try to keep a package file and environment+whatever hacks (like module loads) in sync.

And I was actually fairly happy with that when `spack setup` still existed. The codes I usually work with are cmake-based, and spack setup basically created a (python) script that ran cmake for me with all the options so that it'd find the right (spack-installed/external) deps etc.

Unfortunately, despite me liking it, `spack setup` went away. I'd still like to find a way to replicate that experience. I've tried `dev-build`, though not lately, but basically that didn't suit me, it was already pretty much a showstopper that I didn't have control over the build directory (I really like to have the ability to have multiple builds), but it was also really inconvenient for any kind of incremental change, rebuild, test kind of cycle. I also remember that sourcing the `spack-build-env.txt` pretty much destroyed my shell, pointing me to a non-workable python, what not. Not sure if all that is still the case.

I just tried `spack build-env` again (I can't quite remember what problems I had back in the day). It's actually hard to google how it's supposed to be used. I used `-d env.sh` and then source the environment, and while slightly cumbersome, that actually seems to work fine, so that may be my future go-to ;)

I tried `spack -e build-env <pkg> bash`, and that kind reminded me of the mysteries I used to have:
```
[kaig1@login3 summit-cpu-pkg]$ spack -e . build-env gene bash
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007fffafaa58a0 (most recent call first):
```

Also as I'm trying an alternative way:  :(
```
[kaig1@login3 summit-cpu-pkg]$ spack -e . build-env cmake ~/src/gene-dev
==> Error: [Errno 13] Permission denied
```

Just as a general comment, I do understand that trying to accommodate all kinds of environments, packages, etc, is a total can of worms. But IMHO it'd also really make life easier if introducing yet another layer of complexity (like the spack compiler wrappers and whatever magic is done to the environment) could be avoided, when it's not necessary, like in the cmake case. I'd be so much happier if spack could just make a `cmake.sh` for me which calls cmake with the right compilers, `CMAKE_PREFIX_PATH`, etc. And if something goes wrong, or I need to adapt something (say I want to add a temporary flag to cmake), it's easy to edit, and easy to debug.

For a proper cmake project, one shouldn't really have to introduce another layer of wrappers -- obviously, in many situations it is required to do that to work around limitations in the original build system, but I kinda don't like to suffer myself because others use custom legacy makefiles, so couldn't they only be used where needed (or, at least, be opt-out, ie, if I could say in my package.py: use cmake natively).

**Comment by scheibelp** [comment]
> I've tried dev-build, though not lately, but basically that didn't suit me, it was already pretty much a showstopper that I didn't have control over the build directory (I really like to have the ability to have multiple builds),

Out of curiosity, how do these builds differ? We've spoken to users who do this in different environments with different compilers, and I could also see a desire to rebuild with different variants.

Overall though, we in fact *do* allow multiple builds of the same development source with CMake.

> but it was also really inconvenient for any kind of incremental change, rebuild, test kind of cycle

We definitely want to improve `spack develop` to make this easy, so if you have more comments on pain points here we are interested in addressing them . `spack install` should address the necessary environment setup that `spack setup` originally did (to be clear, I think this is true now). https://github.com/spack/spack/issues/11582#issuecomment-906656438 highlights that the initial setup is a bit tedious but after that the general flow is to repeat steps 10-12.

**Comment by scheibelp** [comment]
> highlights that the initial setup is a bit tedious but after that the general flow is to repeat steps 10-12.

Actually you wouldn't repeat step 10, and if the source changes but the spec doesn't then I think `spack install` should rebuild the specs under development (@becker33 this is based on timestamps right?).

**Comment by scheibelp** [comment]
BTW if you are interested, we have a (currently) weekly meeting on Wednesdays if you think a faster-paced discussion would be useful (e.g. https://groups.google.com/g/spack/c/81AHtT2Gmjw).

**Comment by germasch** [comment]
The problems I had related to build-env are because I missed `--`:

```
spack -e . build-env gene -- cmake [...]
```

Reading the help text helps ;) (But a more helpful error would be nice, too)

**Comment by germasch** [comment]
So that Python_Initialize error wasn't just missing `--`, it just happened again. And if that happens, somehow the `spack.{lock,yaml}` end up moved out of `.spack-env` into `.`, breaking the environment altogether. I've re-created my environment from scratch 10 times in the last 30 minutes as things went awry.

If I'm careful to not do something else, the `spack build-env` approach actually does work fine, but besides being a bit cumbersome, it also still has some real drawback for me:

* Unsurprisingly `make` only works after setting up the environment correctly. This means if I log again the next day, I have to source the environment again. I guess that's not too bad. But it also means, e.g., that integration with an IDE doesn't work, at least not without jumping through hoops. I used to be an emacs+terminal person (and I think that problem also happens when compiling from within emacs), but i guess i've moved over to the dark side and i'm usually using vs code now, and that is usually pretty good at making things work out of the box, but obviously can't do things right automatically here. If I run `cmake` myself, I'll actually end up with things configured in a way that cmake remembers my compilers etc and `make` (or `ninja`) just works.
* I've previously run into other fairly evil problems, e.g., where I used an installed anaconda-python as an external in spack, which made spack add anaconda include and library paths into the build and pick up some incompatible library version of something. And that's another problem in general: spack will compile the code with its own compiler wrappers, which adds all kind of include and link paths. It usually doesn't cause trouble, but it's certainly happened to me, and it's hard to debug since it's all happening opaquely under the hood. OTOH, it also means if i develop using spack I can't rely on my code still building in other environments. For example, as long as the package says I'm using fftw, my code will find the fftw headers, even if i screwed up my `CMakeLists.txt` handling of fftw. Or, where spack things it knows what to put into RPATH and ends up making the code pick up the wrong library.

Sorry about sounding so negative. I actually like spack, and I do use it quite a bit, but pretty much every time I use it I run into non-obvious problems, and it's been even harder to get others to use it.

And sadly, I have another demonstration of that:

The `spack -e . develop` approach described above (which I've not tried before, but again, I'd actually like to find a simple way that suits me) runs into troubles with the same code that works via build-env:

```
[kaig1@login5 summit-cpu-pkg]$ spack -e . install
[...]
     449      315 |        call mpi_bcast(ExB_stime,1,MPI_REAL_TYPE,0,MY_MPI_COMM_WORLD,ierr)
     450          |                      1
  >> 451    Error: Type mismatch between actual argument at (1) and actual argument at (2) (REAL(8)/LOGICAL(4)).
```

I know this error (gfortran-10), but it shouldn't happen since my `CMakeLists.txt` will add `-fallow-argument-mismatch`. Except, apparently, it didn't, but it did when i used build-env? The build directory is of course not where I'd like it to be, but in some path that can definitely not be typed quickly but has to be copy and pasted, and again I'll have to source the environment before I can try to figure out what went wrong.

And just generally, even if it worked, `spack -e install` is a total productivity killer the way I work. I frequently compile code while writing just to find the obvious (compile-time errors), and I often will just build a single unit test and run it, track down what's wrong, rinse and repeat. Waiting 10 minutes for a rebuild vs 5 seconds for building a single test makes that impossible. (Since `spack -e install` doesn't work, I don't know in how far it doesn't rerun cmake etc, but even if just calls `make` again, vs `make <target>`, that can still make a real big difference.

I know I'm being quite particular here, and others may have very different approaches. But that's also part of my point, I'd actually like spack to do less, not change up all kinds of stuff under me. I pretty just would like to have it call cmake for me, corresponding to the variants and tell cmake about the compilers and libs that I want to use, and that's it. Pretty much like what `spack setup` did, and that then leaves the code developer to go whatever way works for them.

**Comment by haampie** [comment]
@germasch you can open a github discussion about improving developer experience, cause this is rather unrelated to the original issue.

short reply: use `spack develop` + `spack install` when you want to test spack installs end-to-end (e.g. you're debugging a dev version of a dependency or so and you want to rebuild dependency + dependent), use `spack build-env` if you want to quickly iterate with make. Maybe we can add an option like `spack install --drop-shell` which opens a shell on failure with working directory set to the staging/build dir, so you can run make until you're happy. Regarding your editor, I don't know, maybe open it from build-env, maybe make it source `spack -e . build-env [pkg name] > env.txt` on startup.

I've already opened a discussion about improving `spack develop` (https://github.com/spack/spack/discussions/24966), and I know that there are people who actually want this stuff to land, it's just that it's a very low-priority issue for core developers, so it'd be better for people who use spack for development to be more vocal and actually engage so that we can raise the priority a bit and get prs merged.
